### DISCUSSION TITLE
Can't override framework configuration loaded by YamlFileLoader inside prependExtension()

### DISCUSSION THREAD

**Discussion by 99hops** [symfony/symfony#47478] [Q&A]
I have a bundle that extends AbstractBundle with prependExtension() function like 
```
public function prependExtension(ContainerConfigurator $container, ContainerBuilder $builder): void
{
        $fileLoader = new YamlFileLoader(
            $builder,
            new FileLocator([__DIR__.'/../config']),
            $builder->getParameter('kernel.environment')
        );

        $fileLoader->load('packages/framework.yaml');
}
```
Bundle packages/framework.yaml
```
framework:
        trusted_proxies: 'a.a.a.a'
```
Main packages/framework.yaml
```
framework:
        trusted_proxies: 'b.b.b.b'
```

Problem is trusted_proxies can't be overridden by main packages/framework.yaml
`bin/console debug:container --parameter=kernel.trusted_proxies`  = `kernel.trusted_proxies   a.a.a.a`

I can if I do this inside prependExtension()
```
$definitions = Yaml::parseFile(__DIR__.'/../config/packages/framework.yaml');
foreach ($definitions as $name => $config) {
   $builder->prependExtensionConfig($name, $config);
}
```
but this doesn't work when there is when@ inside and requires files to be parsed in specific order + import parameters first by $builder->setParameter(), so it's not ideal.

Question is how do I override non-bundle (like framework:) parameters added from bundle within "main" config files under project config/packages yaml files?

**Comment by stof** [comment]
If a bundle loads additional files instead of prepending them, it won't let the project-level config win, as it will be after the project config.

Usages of `prependExtension` are meant to use `prependExtensionConfig`.

**Reply by 99hops** [reply]
@stof yes, this is the case. Is there a way to prepend instead of load yaml file (same as with YamlFileLoader) from a bundle to achieve the desired behavior?

**Reply by stof** [reply]
if you want to use the YamlFileLoader, I would say that you should create a new temporary ContainerBuilder, pass that one to the loader and then read the extension configs from that temporary builder and prepend them in the actual builder being configured.

**Reply by 99hops** [reply]
I took a look and there seems to be no easy way to do this. 
- `ContainerBuilder->merge($tempCb);` doesn't work as expected (it won't let the project-level config win) 
- there is no ContainerBuilder->getExtensionConfigs() method to access `private array $extensionConfigs = [];` to use for `$builder->prependExtensionConfig($name, $config);`

What options do I have to do this? :smiley:

**Comment by 99hops** [comment]
@stof This is working, what do you think?

```
public function prependExtension(ContainerConfigurator $container, ContainerBuilder $builder): void
{
    $cb = new ContainerBuilder($builder->getParameterBag());

     foreach ($builder->getExtensions() as $extension) {
         $cb->registerExtension($extension);
     }

    $fileLoader = new YamlFileLoader(
        $cb,
        new FileLocator([__DIR__.'/../config']),
        $cb->getParameter('kernel.environment')
    );

    $fileLoader->load('packages/framework.yaml');
    $fileLoader->load('packages/monolog.yaml');
    $fileLoader->load('packages/jms_serializer.yaml');
    $fileLoader->load('packages/doctrine.yaml');
    $fileLoader->load('packages/doctrine_migrations.yaml');

    foreach ($cb->getParameterBag()->all() as $param => $value) {
        $builder->setParameter($param, $value);
    }

    foreach ($cb->getExtensions() as $extension) {
        foreach ($cb->getExtensionConfig($extension->getAlias()) as $config) {
            $builder->prependExtensionConfig($extension->getAlias(), $config);
        }
    }
}
```
